larreco/RecoAlg/PMAlg/PmaTrack3D.h

```diff
@@ -0,0 +1,318 @@
+
+	void MakeProjection(void);
+	void UpdateProjection(void);
+	void SortHits(void);
+
+	unsigned int DisableSingleViewEnds(void);
+	bool SelectHits(float fraction = 1.0F);
+	bool SelectRndHits(size_t segmax, size_t vtxmax);
+	bool SelectAllHits(void);
+
+	float GetEndSegWeight(void) const { return fEndSegWeight; }
+	void SetEndSegWeight(float value) { fEndSegWeight = value; }
+
+	float GetPenalty(void) const { return fPenaltyFactor; }
+	void SetPenalty(float value) { fPenaltyFactor = value; }
+
+	unsigned int GetMaxHitsPerSeg(void) const { return fMaxHitsPerSeg; }
+	void SetMaxHitsPerSeg(unsigned int value) { fMaxHitsPerSeg = value; }
+
+	ETag GetTag(void) const { return fTag; }
+	void SetTag(ETag value) { fTag = value; }
+
+private:
+	void ClearNodes(void);
+	void MakeFastProjection(void);
+
+	bool AttachToSameTPC(pma::Node3D* vStart);
+	bool AttachToOtherTPC(pma::Node3D* vStart);
+
+	bool AttachBackToSameTPC(pma::Node3D* vStart);
+	bool AttachBackToOtherTPC(pma::Node3D* vStart);
+
+	void InternalFlip(std::vector< pma::Track3D* >& toSort);
+
+	void UpdateHitsRadius(void);
+	double AverageDist2(void) const;
+
+	bool InitFromHits(int tpc, int cryo, float initEndSegW = 0.05F);
+	bool InitFromRefPoints(int tpc, int cryo);
+	void InitFromMiddle(int tpc, int cryo);
+
+	pma::Track3D* GetNearestTrkInTree(const TVector3& p3d_cm, double& dist, bool skipFirst = false);
+	pma::Track3D* GetNearestTrkInTree(const TVector2& p2d_cm,
+		unsigned int view, unsigned int tpc, unsigned int cryo,
+		double& dist, bool skipFirst = false);
+	void ReassignHitsInTree(pma::Track3D* plRoot = 0);
+
+	/// Distance to the nearest subsequent (dir = Track3D::kForward) or preceeding (dir = Track3D::kBackward)
+	/// hit in given view. In case of last (first) hit in this view the half-distance in opposite direction is
+	/// returned. Parameter secondDir is only for internal protection - please leave the default value.
+	double HitDxByView(size_t index, unsigned int view, Track3D::EDirection dir, bool secondDir = false) const;
+
+	/// Calculate 3D position corresponding to 2D hit, return true if the 3D point is
+	/// in the same TPC as the hit, false otherwise. Calculates also distance^2 between
+	/// the hit and 2D projection of the track. NOTE: results are meaningful only if
+	/// the function returns true.
+	bool GetUnconstrainedProj3D(art::Ptr<recob::Hit> hit, TVector3& p3d, double& dist2) const;
+
+	void RebuildSegments(void);
+	bool SwapVertices(size_t v0, size_t v1);
+	void UpdateParams(void);
+
+	bool CheckEndSegment(pma::Track3D::ETrackEnd endCode);
+
+	int index_of(const pma::Hit3D* hit) const;
+	std::vector< pma::Hit3D* > fHits;
+
+	std::vector< TVector3* > fAssignedPoints;
+
+	pma::Element3D* GetNearestElement(const TVector2& p2d, unsigned int view, int tpc = -1,
+		bool skipFrontVtx = false, bool skipBackVtx = false) const;
+	pma::Element3D* GetNearestElement(const TVector3& p3d) const;
+	std::vector< pma::Node3D* > fNodes;
```

Comment:
We understand that nodes can be shared among tracks.  Suggest that they be held by shared ptr, not bare pointer.